text_generation/causal_lm/cpp/speculative_decoding_lm.cpp

```diff
@@ -0,0 +1,294 @@
+    for (size_t idx = 1; idx < num_outputs; ++idx) {
+        auto old_tensor = request.get_output_tensor(idx);
+        auto trimmed_tensor = trimm_tensor(old_tensor, seq_len_axis, new_seq_len);
+        request.set_output_tensor(idx, trimmed_tensor);
+    }
+}
+
+int main(int argc, char* argv[]) try {
+    if (argc != 4) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <DRAFT MODEL_DIR> <MAIN MODEL_DIR> '<PROMPT>'");
+    }
+
+    // tokenizer model
+    ov::Core core;
+    core.add_extension(OPENVINO_TOKENIZERS_PATH);  // OPENVINO_TOKENIZERS_PATH is defined in CMakeLists.txt
+    // tokenizer and detokenizer work on CPU only
+    ov::InferRequest tokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_tokenizer.xml", "CPU").create_infer_request();
+    auto [draft_input_ids, draft_attention_mask] = tokenize(tokenizer, argv[3]);
+    ov::InferRequest detokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    TextStreamer text_streamer{std::move(detokenizer)};
+
+    // draft model
+    ov::InferRequest draft_model = core.compile_model(std::string{argv[1]} + "/openvino_model.xml", "CPU").create_infer_request();
+
+    draft_model.set_tensor("input_ids", draft_input_ids);
+    draft_model.set_tensor("attention_mask", draft_attention_mask);
+    
+    ov::Tensor draft_position_ids = draft_model.get_tensor("position_ids");
+    draft_position_ids.set_shape(draft_input_ids.get_shape());
+    std::iota(draft_position_ids.data<int64_t>(), draft_position_ids.data<int64_t>() + draft_position_ids.get_size(), 0);
+    uint64_t seq_len = draft_input_ids.get_shape()[1];
+
+    // main model
+    ov::InferRequest main_model = core.compile_model(std::string{argv[2]} + "/openvino_model.xml", "CPU").create_infer_request();
+
+    // Input tensors for the main model should not be mixed with draft.
+    // Do not feed the same draft_postion_ids to the main, but copy input_ids from the draft_input_ids
+    auto input_ids = main_model.get_tensor("input_ids");
+    input_ids.set_shape(draft_input_ids.get_shape());
+    for (int i = 0; i < seq_len; ++i)
+        input_ids.data<int64_t>()[i] = draft_input_ids.data<int64_t>()[i];
+
+    auto attention_mask = main_model.get_tensor("attention_mask");
+    attention_mask.set_shape(draft_input_ids.get_shape());
+    std::fill_n(attention_mask.data<int64_t>(), attention_mask.get_size(), 1);
+
+    auto position_ids = main_model.get_tensor("position_ids");
+    position_ids.set_shape(draft_input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+    
+    // To coollect kv-cache for the <PROMPT> and to get the next token run the very first infer request
+    init_key_values(draft_model);
+    init_key_values(main_model);
+    draft_model.infer();
+    main_model.infer();
```

Comment:
do we need to infer main model before main loop?
I suppose first main_model infer can be done for prompt + K  draft tokens

E.g. we can have do / while cycle instead of iteration before loop + do / while